File: Circular_Linked_List.py
# Menu driven program to show operations in circular linked list :-

class Node:
    def __init__(self, data):
        self.data = data
        self.next = None

class CircularLinkedList:
    def __init__(self):
        self.head = None

    # Insert at the beginning
    def insert_at_beginning(self, data):
        new_node = Node(data)
        if not self.head:
            new_node.next = new_node
            self.head = new_node
        else:
            temp = self.head
            while temp.next != self.head:
                temp = temp.next
            temp.next = new_node
            new_node.next = self.head
            self.head = new_node

    # Insert at the end
    def insert_at_end(self, data):
        new_node = Node(data)
        if not self.head:
            new_node.next = new_node
            self.head = new_node
        else:
            temp = self.head
            while temp.next != self.head:
                temp = temp.next
            temp.next = new_node
            new_node.next = self.head

    # Delete a node
    def delete_node(self, data):
        if not self.head:
            print("List is empty!")
            return

        temp = self.head
        prev = None

        # If the node to be deleted is the head
        if temp.data == data:
            if temp.next == self.head:
                self.head = None
            else:
                while temp.next != self.head:
                    temp = temp.next
                temp.next = self.head.next
                self.head = self.head.next
            print("Node deleted successfully.")
            return

        prev = self.head
        temp = self.head.next

        while temp != self.head:
            if temp.data == data:
                prev.next = temp.next
                print("Node deleted successfully.")
                return
            prev = temp
            temp = temp.next

        print("Node not found!")

    # Display the list
    def display(self):
        if not self.head:
            print("List is empty!")
            return
        temp = self.head
        print("Circular Linked List:", end=" ")
        while True:
            print(temp.data, end=" -> ")
            temp = temp.next
            if temp == self.head:
                break
        print("(Head)")

    # Search for a node
    def search(self, data):
        if not self.head:
            print("List is empty!")
            return
        temp = self.head
        pos = 1
        while True:
            if temp.data == data:
                print(f"Element found at position {pos}")
                return
            temp = temp.next
            pos += 1
            if temp == self.head:
                break
        print("Element not found in the list.")

# Menu-driven interface
cll = CircularLinkedList()
while True:
    print("\nCircular Linked List Operations:")
    print("1. Insert at Beginning\n2. Insert at End\n3. Delete a Node\n4. Display List\n5. Search for a Node\n6. Exit")
    choice = int(input("Enter your choice: "))
    
    if choice == 1:
        value = int(input("Enter value: "))
        cll.insert_at_beginning(value)
    elif choice == 2:
        value = int(input("Enter value: "))
        cll.insert_at_end(value)
    elif choice == 3:
        value = int(input("Enter value to delete: "))
        cll.delete_node(value)
    elif choice == 4:
        cll.display()
    elif choice == 5:
        value = int(input("Enter value to search: "))
        cll.search(value)
    elif choice == 6:
        print("Exiting program...")
        break
    else:
        print("Invalid choice! Try again.")

File: Circular_Linked_List.c
// Menu driven program to show operations in circular linked list :-

#include <stdio.h>
#include <stdlib.h>

// Node structure
struct Node {
    int data;
    struct Node* next;
};

// Head pointer
struct Node* head = NULL;

// Function to insert at the beginning
void insertAtBeginning(int value) {
    struct Node* newNode = (struct Node*)malloc(sizeof(struct Node));
    newNode->data = value;
    if (head == NULL) {
        newNode->next = newNode;
        head = newNode;
    } else {
        struct Node* temp = head;
        while (temp->next != head) {
            temp = temp->next;
        }
        newNode->next = head;
        temp->next = newNode;
        head = newNode;
    }
}

// Function to insert at the end
void insertAtEnd(int value) {
    struct Node* newNode = (struct Node*)malloc(sizeof(struct Node));
    newNode->data = value;
    if (head == NULL) {
        newNode->next = newNode;
        head = newNode;
    } else {
        struct Node* temp = head;
        while (temp->next != head) {
            temp = temp->next;
        }
        temp->next = newNode;
        newNode->next = head;
    }
}

// Function to delete a node
void deleteNode(int value) {
    if (head == NULL) {
        printf("List is empty!\n");
        return;
    }

    struct Node* temp = head;
    struct Node* prev = NULL;

    // If the node to be deleted is the head
    if (head->data == value) {
        if (head->next == head) { 
            free(head);
            head = NULL;
        } else {
            while (temp->next != head) {
                temp = temp->next;
            }
            temp->next = head->next;
            free(head);
            head = temp->next;
        }
        printf("Node deleted successfully.\n");
        return;
    }

    // Searching for the node
    while (temp->next != head && temp->data != value) {
        prev = temp;
        temp = temp->next;
    }

    if (temp->data == value) {
        prev->next = temp->next;
        free(temp);
        printf("Node deleted successfully.\n");
    } else {
        printf("Node not found!\n");
    }
}

// Function to display the list
void display() {
    if (head == NULL) {
        printf("List is empty!\n");
        return;
    }
    struct Node* temp = head;
    printf("Circular Linked List: ");
    do {
        printf("%d -> ", temp->data);
        temp = temp->next;
    } while (temp != head);
    printf("(Head)\n");
}

// Function to search for a node
void search(int value) {
    if (head == NULL) {
        printf("List is empty!\n");
        return;
    }
    struct Node* temp = head;
    int pos = 1;
    do {
        if (temp->data == value) {
            printf("Element found at position %d\n", pos);
            return;
        }
        temp = temp->next;
        pos++;
    } while (temp != head);
    printf("Element not found in the list.\n");
}

// Menu-driven main function
int main() {
    int choice, value;
    while (1) {
        printf("\nCircular Linked List Operations:\n");
        printf("1. Insert at Beginning\n2. Insert at End\n3. Delete a Node\n4. Display List\n5. Search for a Node\n6. Exit\n");
        printf("Enter your choice: ");
        scanf("%d", &choice);
        switch (choice) {
            case 1:
                printf("Enter value: ");
                scanf("%d", &value);
                insertAtBeginning(value);
                break;
            case 2:
                printf("Enter value: ");
                scanf("%d", &value);
                insertAtEnd(value);
                break;
            case 3:
                printf("Enter value to delete: ");
                scanf("%d", &value);
                deleteNode(value);
                break;
            case 4:
                display();
                break;
            case 5:
                printf("Enter value to search: ");
                scanf("%d", &value);
                search(value);
                break;
            case 6:
                exit(0);
            default:
                printf("Invalid choice! Try again.\n");
        }
    }
    return 0;
}
